src/messages_manager.ts

```diff
@@ -20,38 +21,30 @@ export class MessagesManager implements Disposable {
         disposeAll(this.disposables);
     }
 
-    private handleRedraw({ name, args }: EventBusData<"redraw">): void {
+    private async handleRedraw({ name, args }: EventBusData<"redraw">): Promise<void> {
         switch (name) {
             case "msg_show": {
-                const lines: string[] = [];
-                for (const [type, content, clear] of args) {
-                    if (type === "return_prompt") {
-                        // Ignore return_prompt
-                        //
-                        // A note to future readers: return_prompt is sent much more often with ui_messages. It may
-                        // not do what you expect from what :help ui says, so be careful about using these events.
-                        // See: https://github.com/vscode-neovim/vscode-neovim/issues/2046#issuecomment-2144175058
-                        continue;
-                    }
-
-                    if (clear) {
-                        // Remove all stored lines, we will clear the output console on the way out.
-                        lines.splice(0, lines.length);
-                    }
-
-                    const segments = content.map((c) => c[1]);
-                    if (segments.length > 0) {
-                        segments[0] = segments[0].replace(/^\n+/, "");
-                    }
-
-                    lines.push(...segments);
-                }
-
-                if (lines.length >= 2) {
+                const hasReturnPrompt = args.some(([type]) => type === "return_prompt");
+                const msg = args.reduce((acc, [type, content, clear]) => {
+                    // Ignore return_prompt
+                    //
+                    // A note to future readers: return_prompt is sent much more often with ui_messages. It may
+                    // not do what you expect from what :help ui says, so be careful about using these events.
+                    // See: https://github.com/vscode-neovim/vscode-neovim/issues/2046#issuecomment-2144175058
+                    if (type === "return_prompt") return acc;
+                    if (clear) return "";
+                    return acc + content.map((c) => c[1]).join("");
+                }, "");
+                const outputMsg = hasReturnPrompt ? msg.replace(/\n$/, "") : msg;
```

Comment:
I know this was like this, but we may consider trimming trailing all newlines from here (`\n*$`) to minimize accidental pop outs

Reply:
However, this will cause the expected line breaks to be removed.

Reply:
It could. But if they're at the end of the string, that likely doesn't matter.